File: ios/GradeThread/Disclosure/DisclosureService.swift
import Foundation

/// Defect-disclosure edge client. Thin wrapper over `EdgeAPI.shared` (auto-auth,
/// snake_case ⇄ camelCase). Endpoints already exist on the Hono service and are
/// mounted at `/api/flipdesk/disclosure/*`. Behind a protocol so the store is
/// unit-testable with an in-memory mock (mirrors `AspectsProviding`).
protocol DisclosureProviding {
    func disclosure(itemId: String) async throws -> DisclosureData
    func saveAnnotated(itemId: String, imageType: String, dataURL: String) async throws -> SaveAnnotatedResponse
    func applyToListing(itemId: String) async throws -> ApplyDisclosureResponse
}

struct DisclosureService: DisclosureProviding {
    private let api: EdgeAPI

    init(api: EdgeAPI = .shared) {
        self.api = api
    }

    private func base(_ itemId: String) -> String {
        "/api/flipdesk/disclosure/item/\(itemId)"
    }

    func disclosure(itemId: String) async throws -> DisclosureData {
        try await api.getJSON(base(itemId))
    }

    func saveAnnotated(itemId: String, imageType: String, dataURL: String) async throws -> SaveAnnotatedResponse {
        // imageType → image_type, dataUrl → data_url via the shared encoder.
        struct Body: Encodable {
            let imageType: String
            let dataUrl: String
        }
        return try await api.postJSON(
            "\(base(itemId))/annotated-photo",
            body: Body(imageType: imageType, dataUrl: dataURL)
        )
    }

    func applyToListing(itemId: String) async throws -> ApplyDisclosureResponse {
        struct Empty: Encodable {}
        return try await api.postJSON("\(base(itemId))/apply-to-listing", body: Empty())
    }
}

File: android/app/src/main/java/com/gradethread/app/disclosure/DisclosureService.kt
package com.gradethread.app.disclosure

import com.gradethread.app.platform.net.EdgeApi
import com.gradethread.app.platform.net.EdgeApiError
import javax.inject.Inject
import javax.inject.Named
import javax.inject.Singleton

/**
 * US-1360: the disclosure endpoints.
 *
 * `apply-to-listing` edits a LIVE eBay description, so its result is reported
 * verbatim rather than assumed — see [DisclosureViewModel].
 */
@Singleton
class DisclosureService @Inject constructor(
    @Named("shared") private val edge: EdgeApi,
) {

    companion object {
        fun itemPath(itemId: String) = "/api/flipdesk/disclosure/item/$itemId"
        fun annotatedPhotoPath(itemId: String) = "${itemPath(itemId)}/annotated-photo"
        fun applyPath(itemId: String) = "${itemPath(itemId)}/apply-to-listing"
    }

    suspend fun disclosure(itemId: String): DisclosureData = edge.json.decodeFromString(
        DisclosureData.serializer(),
        edge.getRaw(itemPath(itemId)),
    )

    /**
     * Upload one annotated composite.
     *
     * [dataUrl] must be a base64 PNG data URL — the server sniffs the real
     * bytes and refuses anything else, because this lands in the PUBLIC bucket
     * where an SVG smuggled behind a png prefix would be served as stored.
     */
    suspend fun saveAnnotated(
        itemId: String,
        imageType: String,
        dataUrl: String,
    ): SaveAnnotatedResponse = edge.json.decodeFromString(
        SaveAnnotatedResponse.serializer(),
        edge.postRaw(
            annotatedPhotoPath(itemId),
            edge.json.encodeToString(
                AnnotatedPhotoRequest.serializer(),
                AnnotatedPhotoRequest(imageType, dataUrl),
            ),
        ),
    )

    /** Push the disclosure text into the live listing's description. */
    suspend fun applyToListing(itemId: String): ApplyDisclosureResponse =
        edge.json.decodeFromString(
            ApplyDisclosureResponse.serializer(),
            edge.postRaw(applyPath(itemId), "{}"),
        )

    fun message(error: Throwable): String =
        (error as? EdgeApiError)?.userMessage()
            ?: error.message
            ?: "Couldn't reach the disclosure service."
}
